diff -ur cluster-2.00.00/gfs2/mkfs/Makefile cluster-2.00.00-o/gfs2/mkfs/Makefile
--- cluster-2.00.00/gfs2/mkfs/Makefile	2007-04-05 14:34:25.000000000 -0600
+++ cluster-2.00.00-o/gfs2/mkfs/Makefile	2006-08-22 13:16:39.000000000 -0600
@@ -35,7 +35,7 @@
 
 install: all
 	${INSTALL} -m 0755 ${MKFS} ${sbindir}
-	ln -f ${MKFS} ${sbindir}/${JADD}
+	ln -s ${MKFS} ${sbindir}/${JADD}
 
 clean:
 	rm -f *.o ${MKFS} ${JADD}
